solidity/contracts/AbstractBondedECDSAKeep.sol

```diff
@@ -0,0 +1,643 @@
+        return status == Status.Closed;
+    }
+
+    /// @notice Returns true if the keep has been terminated.
+    /// Keep is terminated when bonds are seized and members no longer support
+    /// this keep.
+    /// @return true if the keep has been terminated, false otherwise.
+    function isTerminated() public view returns (bool) {
+        return status == Status.Terminated;
+    }
+
+    /// @notice Returns members of the keep.
+    /// @return List of the keep members' addresses.
+    function getMembers() public view returns (address[] memory) {
+        return members;
+    }
+
+    /// @notice Checks a fraud proof for a valid signature from this keep that was
+    /// not first approved via a call to sign.
+    /// @dev The function expects the signed digest to be calculated as a sha256 hash
+    /// of the preimage: `sha256(_preimage))`. The digest is verified against the
+    /// preimage to ensure the security of the ECDSA protocol. Verifying just the
+    /// signature and the digest is not enough and leaves the possibility of the
+    /// the existential forgery. If digest and preimage verification fails the
+    /// function reverts.
+    /// Reverts if a public key has not been set for the keep yet.
+    /// @param _v Signature's header byte: `27 + recoveryID`.
+    /// @param _r R part of ECDSA signature.
+    /// @param _s S part of ECDSA signature.
+    /// @param _signedDigest Digest for the provided signature. Result of hashing
+    /// the preimage with sha256.
+    /// @param _preimage Preimage of the hashed message.
+    /// @return True if fraud, false otherwise.
+    function checkSignatureFraud(
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s,
+        bytes32 _signedDigest,
+        bytes memory _preimage
+    ) public view returns (bool _isFraud) {
+        require(publicKey.length != 0, "Public key was not set yet");
+
+        bytes32 calculatedDigest = sha256(_preimage);
+        require(
+            _signedDigest == calculatedDigest,
+            "Signed digest does not match sha256 hash of the preimage"
+        );
+
+        bool isSignatureValid = publicKeyToAddress(publicKey) ==
+            ecrecover(_signedDigest, _v, _r, _s);
+
+        // Check if the signature is valid but was not requested.
+        return isSignatureValid && digests[_signedDigest] == 0;
+    }
+
+    /// @notice Initialization function.
+    /// @dev We use clone factory to create new keep. That is why this contract
+    /// doesn't have a constructor. We provide keep parameters for each instance
+    /// function after cloning instances from the master contract.
+    /// @param _owner Address of the keep owner.
+    /// @param _members Addresses of the keep members.
+    /// @param _honestThreshold Minimum number of honest keep members.
+    function initializeBondedECDSAKeep(
+        address _owner,
+        address[] memory _members,
+        uint256 _honestThreshold,
+        address _bonding
```

Comment:
Can we name this parameter `_keepBonding` or rename `_keepBonding` parameter in `BondedECDSAKeep` to `_bonding`.

Reply:
I named it this way as `KeepBonding` contract is specific for current bonding with token staking. For ETH only bonding we won't use `KeepBonding` contract, hence the name `_keepBonding` won't be accurate in abstract contract. In `BondedECDSAKeep` contract we are using `KeepBonding` contract, so the parameter name there is `_keepBonding`.

If you want to have both aligned I can agree on using `_bonding`.

Reply:
We name the contract `KeepBonding` because it's Keep-org bonding contract, not because we are bonding KEEP there. I am fine with just `_bonding`.

Reply:
Then we should rename `KeepBonding` contract as it's strictly related to KEEP by token ownership, grants and `TokenStaking`. For ETH only staking we will still be in Keep-org land but `KeepBonding` couldn't be used. A better name for current `KeepBonding` would be `TokenBonding` (because of the relation to `TokenStaking` contract).

Reply:
Ah... I see what you mean. `KeepBonding` is no longer just about organization prefix but about the fact we refer there to `TokenStaking` and `TokenGrant` and we'll not refer to them for ETH-only version. Cool, I think the original naming convention makes a lot of sense then.